File: demo/csharp/MessageDemo.cs
// mq9 C# SDK — Message Demo
//
// Demonstrates:
//  1. Create a mailbox
//  2. Send messages with different priorities
//  3. Fetch + ACK (stateful consumption)
//  4. Consume loop (auto poll)
//  5. Message attributes: key dedup, tags, delay, ttl
//  6. Query without affecting offset
//  7. Delete a message

using System.Text;
using System.Text.Json;
using Mq9;

internal static class MessageDemo
{
    private static readonly string Server =
        Environment.GetEnvironmentVariable("MQ9_SERVER") ?? "nats://demo.robustmq.com:4222";

    public static async Task RunAsync()
    {
        await using var client = new Mq9Client(Server);
        await client.ConnectAsync();

        // ── 1. Create a mailbox ──────────────────────────────────────────
        var address = await client.MailboxCreateAsync(name: "demo.csharp.message", ttl: 300);
        Console.WriteLine($"[mailbox] created: {address}");

        // ── 2. Send messages with different priorities ───────────────────
        var mid1 = await client.SendAsync(address, Json(new { type = "task", id = 1 }));
        Console.WriteLine($"[send] normal    msg_id={mid1}");

        var mid2 = await client.SendAsync(address, Json(new { type = "interrupt", id = 2 }),
            new SendOptions { Priority = Priority.Urgent });
        Console.WriteLine($"[send] urgent    msg_id={mid2}");

        var mid3 = await client.SendAsync(address, Json(new { type = "abort", id = 3 }),
            new SendOptions { Priority = Priority.Critical });
        Console.WriteLine($"[send] critical  msg_id={mid3}");

        // ── 3. Message attributes ────────────────────────────────────────
        // Key dedup: only the latest message with key="status" is kept
        await client.SendAsync(address, Json(new { status = "running" }), new SendOptions { Key = "status" });
        await client.SendAsync(address, Json(new { status = "60%"     }), new SendOptions { Key = "status" });
        var midStatus = await client.SendAsync(address, Json(new { status = "done" }),
            new SendOptions { Key = "status" });
        Console.WriteLine($"[send] dedup key=status, latest msg_id={midStatus}");

        await client.SendAsync(address, Json(new { order = "o-001" }),
            new SendOptions { Tags = ["billing", "vip"] });
        Console.WriteLine("[send] with tags billing,vip");

        await client.SendAsync(address, Json(new { temp = true }), new SendOptions { Ttl = 10 });
        Console.WriteLine("[send] with message ttl=10s");

        var delayedId = await client.SendAsync(address, Json(new { delayed = true }),
            new SendOptions { Delay = 5 });
        Console.WriteLine($"[send] delay=5s  msg_id={delayedId} (returns -1 for delayed)");

        // ── 4. Fetch + ACK (stateful) ────────────────────────────────────
        var messages = await client.FetchAsync(address,
            new FetchOptions { GroupName = "workers", Deliver = "earliest", NumMsgs = 10 });

        Console.WriteLine($"\n[fetch] got {messages.Count} messages (priority order):");
        foreach (var msg in messages)
            Console.WriteLine($"  msg_id={msg.MsgId}  priority={msg.Priority}  payload={Text(msg.Payload)}");

        if (messages.Count > 0)
        {
            await client.AckAsync(address, "workers", messages[^1].MsgId);
            Console.WriteLine($"[ack]   advanced offset to msg_id={messages[^1].MsgId}");
        }

        // ── 5. Query without affecting offset ────────────────────────────
        var results = await client.QueryAsync(address, key: "status");
        Console.WriteLine($"\n[query] key=status → {results.Count} message(s)");
        foreach (var msg in results)
            Console.WriteLine($"  msg_id={msg.MsgId}  payload={Text(msg.Payload)}");

        // ── 6. Consume loop ──────────────────────────────────────────────
        Console.WriteLine("\n[consume] starting loop for 3 s …");

        var consumer = await client.ConsumeAsync(address, async msg =>
        {
            Console.WriteLine($"  [handler] msg_id={msg.MsgId}  payload={Text(msg.Payload)}");
            await Task.CompletedTask;
        }, new ConsumeOptions
        {
            GroupName    = "consume-workers",
            Deliver      = "earliest",
            AutoAck      = true,
            ErrorHandler = async (msg, ex) =>
            {
                Console.Error.WriteLine($"  [error] msg_id={msg.MsgId}  error={ex.Message}");
                await Task.CompletedTask;
            },
        });

        await Task.Delay(3_000);
        await consumer.StopAsync();
        Console.WriteLine($"[consume] stopped. processed={consumer.ProcessedCount}");

        // ── 7. Delete a message ──────────────────────────────────────────
        await client.DeleteAsync(address, mid1);
        Console.WriteLine($"\n[delete] msg_id={mid1} deleted");
    }

    private static byte[] Json(object obj) => JsonSerializer.SerializeToUtf8Bytes(obj);
    private static string Text(byte[] b)   => Encoding.UTF8.GetString(b);
}

File: demo/java/MessageDemo.java
/**
 * mq9 Java SDK — Message Demo
 *
 * Demonstrates:
 *   1. Create a mailbox
 *   2. Send messages with different priorities
 *   3. Fetch + ACK (stateful consumption)
 *   4. Consume loop (auto poll)
 *   5. Message attributes: key dedup, tags, delay, ttl
 *   6. Query without affecting offset
 *   7. Delete a message
 *
 * Compile & run:
 *   mvn compile exec:java -Dexec.mainClass=MessageDemo
 */

import io.mq9.*;
import java.util.List;
import java.util.concurrent.CompletableFuture;

public class MessageDemo {

    private static final String SERVER = "nats://demo.robustmq.com:4222";

    public static void main(String[] args) throws Exception {
        Mq9Client client = Mq9Client.connect(SERVER).get();

        try {
            // ── 1. Create a mailbox ──────────────────────────────────────
            String address = client.mailboxCreate("demo.java.message", 300).get();
            System.out.println("[mailbox] created: " + address);

            // ── 2. Send messages with different priorities ───────────────
            long mid1 = client.send(address, "{\"type\":\"task\",\"id\":1}".getBytes(),
                SendOptions.builder().build()).get();
            System.out.println("[send] normal    msg_id=" + mid1);

            long mid2 = client.send(address, "{\"type\":\"interrupt\",\"id\":2}".getBytes(),
                SendOptions.builder().priority(Priority.URGENT).build()).get();
            System.out.println("[send] urgent    msg_id=" + mid2);

            long mid3 = client.send(address, "{\"type\":\"abort\",\"id\":3}".getBytes(),
                SendOptions.builder().priority(Priority.CRITICAL).build()).get();
            System.out.println("[send] critical  msg_id=" + mid3);

            // ── 3. Message attributes ────────────────────────────────────
            // Key dedup: only the latest message with key="status" is kept
            client.send(address, "{\"status\":\"running\"}".getBytes(),
                SendOptions.builder().key("status").build()).get();
            client.send(address, "{\"status\":\"60%\"}".getBytes(),
                SendOptions.builder().key("status").build()).get();
            long midStatus = client.send(address, "{\"status\":\"done\"}".getBytes(),
                SendOptions.builder().key("status").build()).get();
            System.out.println("[send] dedup key=status, latest msg_id=" + midStatus);

            // Tags
            client.send(address, "{\"order\":\"o-001\"}".getBytes(),
                SendOptions.builder().tags(List.of("billing", "vip")).build()).get();
            System.out.println("[send] with tags billing,vip");

            // Per-message TTL
            client.send(address, "{\"temp\":true}".getBytes(),
                SendOptions.builder().ttl(10L).build()).get();
            System.out.println("[send] with message ttl=10s");

            // Delayed delivery
            long delayedId = client.send(address, "{\"delayed\":true}".getBytes(),
                SendOptions.builder().delay(5L).build()).get();
            System.out.println("[send] delay=5s  msg_id=" + delayedId + " (returns -1 for delayed)");

            // ── 4. Fetch + ACK (stateful) ────────────────────────────────
            List<Message> messages = client.fetch(address, FetchOptions.builder()
                .groupName("workers")
                .deliver("earliest")
                .numMsgs(10)
                .build()).get();
            System.out.println("\n[fetch] got " + messages.size() + " messages (priority order):");
            for (Message msg : messages) {
                System.out.printf("  msg_id=%d  priority=%s  payload=%s%n",
                    msg.msgId, msg.priority, new String(msg.payload));
            }

            if (!messages.isEmpty()) {
                long lastId = messages.get(messages.size() - 1).msgId;
                client.ack(address, "workers", lastId).get();
                System.out.println("[ack]   advanced offset to msg_id=" + lastId);
            }

            // ── 5. Query without affecting offset ────────────────────────
            List<Message> results = client.query(address, "status", null, null).get();
            System.out.println("\n[query] key=status → " + results.size() + " message(s)");
            for (Message msg : results) {
                System.out.printf("  msg_id=%d  payload=%s%n", msg.msgId, new String(msg.payload));
            }

            // ── 6. Consume loop ──────────────────────────────────────────
            System.out.println("\n[consume] starting loop for 3 s …");

            Consumer consumer = client.consume(address, msg -> {
                System.out.printf("  [handler] msg_id=%d  priority=%s  payload=%s%n",
                    msg.msgId, msg.priority, new String(msg.payload));
                return CompletableFuture.completedFuture(null);
            }, ConsumeOptions.builder()
                .groupName("consume-workers")
                .deliver("earliest")
                .autoAck(true)
                .errorHandler((msg, err) ->
                    System.err.printf("  [error]   msg_id=%d  error=%s%n", msg.msgId, err.getMessage()))
                .build()).get();

            Thread.sleep(3000);
            consumer.stop().get();
            System.out.println("[consume] stopped. processed=" + consumer.getProcessedCount());

            // ── 7. Delete a message ──────────────────────────────────────
            client.delete(address, mid1).get();
            System.out.println("\n[delete] msg_id=" + mid1 + " deleted");

        } finally {
            client.close();
        }
    }
}
